contracts/cw1-whitelist/src/contract.rs

```diff
@@ -100,6 +129,49 @@ mod tests {
         );
     }
 
+    #[test]
+    fn binary_serialize_exec() {
+        let original_msg = ExecMsg::Cw1WhitelistContract(ImplExecMsg::RemoveMember {
+            member: "some_member".to_owned(),
+        });
+
+        let serialized_msg = to_binary(&original_msg).unwrap();
+        let serialized_msg: ExecMsg = from_binary(&serialized_msg).unwrap();
+
+        assert_eq!(serialized_msg, original_msg);
+    }
+
+    #[test]
+    fn slice_deserialize_exec() {
+        let deserialized: ExecMsg =
+            from_slice(br#"{"remove_member": {"member": "some_member"}}"#).unwrap();
+        assert_eq!(
+            deserialized,
+            ExecMsg::Cw1WhitelistContract(ImplExecMsg::RemoveMember {
+                member: "some_member".to_owned()
+            })
+        );
+    }
+
+    #[test]
+    fn binary_serialize_query() {
+        let original_msg = QueryMsg::Cw1WhitelistContract(ImplQueryMsg::MembersCnt {});
```

Comment:
Wondering if the JsonSchema is properly generated for this type?
Does it show up flattened or nested?

I think there is no schema.rs file yet (or is that auto-generated)? But something to keep in mind, that it must also flatten these nested structs.

Reply:
Also, using this is code could be annoying if you must write for many test cases. I understand the limits of codegen, but have two possible ideas, maybe one is feasible:

* `QueryMsg::MembersCnt{}` (actually flatten in the proc macros - likely hard)
* `ImplQueryMsg::MembersCnt {}.into()` (just add a helper to wrap it into QueryMsg)

Both are much less typing. Open for other ideas

Reply:
Actually good point, we didn't check it. However I expect it does, as we are adding `#[serde(untagged)]` for proper serialization, and `schemars` should read all serde tags. However - will double-check.

For executing query in real code, you don't use `QueryMsg::Cw1WhitelistContract(ImplQueryMsg::MembersCnt {})` anyway, it is for testing purposes to ensure uniform serialization. Normally you would use `ImplQueryMsg::MembersCnt {}` as its serialization is guaranteed to be the same (and you are working on JSON API-wise). We would prefer to go with `QueryMsg::MembersCnt{}` instead. but here is a problem - it doesn't go well with `#[serde(untagged)]`. We are figuring out the best namings scheme - one of ideas is to rename the final contract messages (as they should almost never be used) to some `GlueQueryMsg` and the contract-specific to just `QueryMsg` so it is very natural call (in this very test it would be `GlueQueryMsg::Cw1WhitelistContract(QueryMsg::MembersCnt {})`, but in real-live query just `QueryMsg::MembersCnt {}`).